Add Creatio Connector docs (#6396)
wso2/docs-apim: en/docs/reference/connectors/creatio-connector/creatio-connector-reference.md

# Creatio Connector Reference

## Authentication implementation

The Creatio connector uses Forms authentication (cookie-based) to authenticate requests with CreatioCRM and the required cookies are retrieved through basic authentication. The Authentication request is created based on the information including the CreatioCRM user credentials configured in the "init" operation of the connector. Once the Creatio REST API provides the authentication response, it is processed by the connector to derive the cookie values.

> The authentication response contains two auth cookies (“BPMCSRF” and “Cookie”) as HTTP header values. 

Reviewer: ```suggestion
The authentication response contains two Auth cookies (“BPMCSRF” and “Cookie”) as HTTP header values. 
```

> The REST API calls related to CreatioCRM CRUD operations are expected to have the above cookies as HTTP headers. Therefore, the connector will validate if the authentication headers present in the incoming API request.

Reviewer: ```suggestion
The REST API calls related to CreatioCRM CRUD operations are expected to have the above cookies as HTTP headers. Therefore, the connector will validate if the authentication headers are present in the incoming API request.
```

If the authentication headers are present in the API request, the connector will send the original request as it is.  

If the HTTP status of the response is 401, 403, 301, and 302 (invalid header value or redirection to regenerate the headers), the connector sends the authentication request again and sets the Auth response header back to the original request and perform the original request.

If the HTTP status of the response is not 401, 403, 301, and 302, then the response is returned to the client app.

> The reason to retry the API calls when giving HTTP status 401, 403, 301, and 302 by Creatio, it returns those HTTP Status when Authentication fails due to multiple reasons and need to retry one time to succeed the relevant API calls. 

Reviewer: Consider rewording

If the headers are not present, the connector sends the authentication request intermediately before the relevant REST API call. After getting the relevant cookie headers, it sends the original request to the Creatio endpoint with the retrieved cookies set as headers and sends back the response.

The authentication request is sent to the Creatio platform and the received response is processed to derive the cookie values that are CSRF token and Cookie header values that contain the necessary auth data for authentication at the Creatio end. These HTTP header values need to be set by the client application when calling any other REST API resource from the Creatio end. After expiry, you will be responsible for getting new CSRF cookie values. 

To use the Creatio connector, add the `<creatiocrm.init>` element in your configuration before carrying out any other Creatio operations.

## Operations

The following operations allow you to work with the Documentum Connector. Click an operation name to see parameter details and samples on how to use it.

### Initialize the connector

??? note "Init"

>     This operation is to create an authentication channel after successful authentication, it will respond back with two main cookie values by the WSO2 EI in-order to use subsequent API requests. 

Reviewer: Is it ok to mention WSO2 EI here?

>     The init operation has the below configurations that are required to obtain the authentication cookies information for the rest of the operations.

Reviewer: ```suggestion
    The init operation has the following configurations that are required to obtain the information of the authentication cookies for the rest of the operations.
```

>     |parameter Name|Description|Required|Sample Value|

Reviewer: ```suggestion
    |Parameter Name|Description|Required|Sample Value|
```

    |:----|:----|:----|:----|
    |hostName|Hostname of the CreatioCRM|Yes|116914-crm-bundle.creatio.com|

>     |password|Password for login to CreatioCRM|Yes|Supervisor|

Reviewer: ```suggestion
    |password|Password to login to CreatioCRM|Yes|Supervisor|
```

>     |username|Username for login to CreatioCRM|Yes|Supervisor|

Reviewer: ```suggestion
    |username|Username to login to CreatioCRM|Yes|Supervisor|
```

    |timeout|Timeout value of the Creatio backend|No|6000|

    **Sample Configuration**

    ```xml

    <creatiocrm.init>
        <hostName>https://<hostname>/</hostName>
        <password>password</password>
        <username>username<username>
    </creatiocrm.init>
    <creatiocrm.authSession/>
    
    ```

    This explains how the API request for the init operation should be and the response coming back from the Creatio connector for the Authentication operation.

|Request Header Field|Type|Description|
|:----|:----|:----|
|Content-Type|String|application/json|

> |Response Header Field|Type|value|Description|

Reviewer: ```suggestion
|Response Header Field|Type|Value|Description|
```

### Contact operations

??? note "Create Contact"
    The REST API of Contact creation operation is triggered by the creatiocrm.contactCreate mediator to create a new Contact with a unique id in the CreatioCRM. 
    
    The message body for the contact creation will be taken from the payload of the existing message context.

    |Field|Type|Description|Remarks|
    |:----|:----|:----|:----|
    |Content-Type|String|application/json|	|
    |BPMCSRF|String|The CSRF Token from the Auth response|E.g., bHZfVOIO4.iSuXwMebGibes|
    |Set-cookie|String|Value of Username cookie from the Auth response|E.g., UserName=83|117|112|101|114|118|105|115|111|114; expires=Sat, 07-May-2022 08:46:12 GMT; path=/; secure; HttpOnly|
    |Cookie|String|Cookie Value from the Auth response|	|

>     **Sample Configuration**

Reviewer: ```suggestion
    **Sample configuration**
```

        "ContactPhoto@odata.mediaEditLink": "Contact(12d21632-8b20-45d9-942f-20fcbbf2957c)/ContactPhoto",
        "ContactPhoto@odata.mediaReadLink": "Contact(12d21632-8b20-45d9-942f-20fcbbf2957c)/ContactPhoto",
        "ContactPhoto@odata.mediaContentType": "application/octet-stream",
        "TwitterAFDAId": "00000000-0000-0000-0000-000000000000",
        "FacebookAFDAId": "00000000-0000-0000-0000-000000000000",
        "LinkedInAFDAId": "00000000-0000-0000-0000-000000000000",
        "PhotoId": "00000000-0000-0000-0000-000000000000",
        "GPSN": "",
        "GPSE": "",
        "Surname": "Test",
        "GivenName": "API",
        "MiddleName": "",
        "Confirmed": true,
        "IsNonActualEmail": false,
        "Completeness": 30,
        "LanguageId": "6ebc31fa-ee6c-48e9-81bf-8003ac03b019",
        "Age": 19,
        "IsEmailConfirmed": false
    }
    ```

??? note "Get Contact"
    The contactGet operation is used to retrieve existing contact details from CreatioCRM with the provided ID. If the given ID does not exist in the CreatioCRM, this will return 404 HTTP status code.

    |Field|Type| |Description|Remarks|
    |:----|:----|:----|:----|:----|

>     |BPMCSRF|String| |The CSRF Token from the Auth response|Ex: bHZfVOIO4.iSuXwMebGibes|

Reviewer: ```suggestion
    |BPMCSRF|String| |The CSRF Token from the Auth response|e.g., bHZfVOIO4.iSuXwMebGibes|
```

>     |Set-cookie|String| |Value of Username cookie from the Auth response|Ex: UserName=83|117|112|101|114|118|105|115|111|114; expires=Sat, 07-May-2022 08:46:12 GMT; path=/; secure; HttpOnly|

Reviewer: ```suggestion
    |Set-cookie|String| |Value of Username cookie from the Auth response|e.g., UserName=83|117|112|101|114|118|105|115|111|114; expires=Sat, 07-May-2022 08:46:12 GMT; path=/; secure; HttpOnly|
```

    |Cookie|String| |Cookie Value from the Auth response| |

    **Sample configuration**

    ```xml

    <creatiocrm.init>
        <hostName>https://creatio-hostname</hostName>
        <password>creatio-username</password>
        <username>creatio-password</username>
    </creatiocrm.init>
    <creatiocrm.contactGet>
        <id>{$ctx:id}</id>
    </creatiocrm.contactGet>

    ```

>     **Sample Response**

Reviewer: ```suggestion
    **Sample response**
```

        "DoNotUseFax": false,
        "DoNotUseSms": false,
        "DoNotUseMail": false,
        "Notes": "",
        "Facebook": "",
        "LinkedIn": "",
        "Twitter": "",
        "FacebookId": "",
        "LinkedInId": "",
        "TwitterId": "",
        "ContactPhoto@odata.mediaEditLink": "Contact(12d21632-8b20-45d9-942f-20fcbbf2957c)/ContactPhoto",
        "ContactPhoto@odata.mediaReadLink": "Contact(12d21632-8b20-45d9-942f-20fcbbf2957c)/ContactPhoto",
        "ContactPhoto@odata.mediaContentType": "application/octet-stream",
        "TwitterAFDAId": "00000000-0000-0000-0000-000000000000",
        "FacebookAFDAId": "00000000-0000-0000-0000-000000000000",
        "LinkedInAFDAId": "00000000-0000-0000-0000-000000000000",
        "PhotoId": "00000000-0000-0000-0000-000000000000",
        "GPSN": "",
        "GPSE": "",
        "Surname": "Test",
        "GivenName": "API",
        "MiddleName": "",
        "Confirmed": true,
        "IsNonActualEmail": false,
        "Completeness": 30,
        "LanguageId": "6ebc31fa-ee6c-48e9-81bf-8003ac03b019",
        "Age": 19,
        "IsEmailConfirmed": false
    }

    ```
    
??? note "Update Contact"

>     The contactUpdate operation is  used to update an existing contact details in the CreatioCRM. 

Reviewer: ```suggestion
    The contactUpdate operation is used to update an existing contact details in the CreatioCRM. 
```

>     If the given ID does not exist in the CreatioCRM, this will create a  new contact with the given ID. The message body for the contact update will be taken from the payload of existing message context.

Reviewer: ```suggestion
    If the given ID does not exist in the CreatioCRM, this will create a new contact with the given ID. The message body for the contact update will be taken from the payload of existing message context.
```

        "Phone": "",
        "MobilePhone": "+1 213 566 34 22",
        "Email": "test@gmail",
        "Completeness": 30,
        "Age": 19
    }

    ```
    
??? note "Delete Contact"
    The contactDelete operation is used to delete a contact that already exists in the CreatioCRM with given unique ID.

    If the given ID does not exist in the CreatioCRM this will return 404 HTTP status code 

    |Field|Type| |Description|Remarks|
    |:----|:----|:----|:----|:----|
    |BPMCSRF|String| |The CSRF Token from the Auth response|E.g., bHZfVOIO4.iSuXwMebGibes|
    |Set-cookie|String| |Value of Username cookie from the Auth response|E.g., UserName=83|117|112|101|114|118|105|115|111|114; expires=Sat, 07-May-2022 08:46:12 GMT; path=/; secure; HttpOnly|
    |Cookie|String| |Cookie Value from the Auth response| |

    **Sample configuration**

    ```xml

    <creatiocrm.init>
        <hostName>https://creatio-hostname</hostName>
        <password>creatio-username</password>
        <username>creatio-password</username>
    </creatiocrm.init>
    <creatiocrm.contactDelete>
        <id>{$ctx:id}</id>
    </creatiocrm.contactDelete>

    ```

### Case operations

??? note "Create Case"

>     Creatio caseCreate operation used in EI mediation to create a new case with a unique id in the CreatioCRM. 

Reviewer: ```suggestion
    Creatio caseCreate operation used in EI mediation to create a new case with a unique ID in the CreatioCRM. 
```

    The message body for the case creation will be taken from the payload of the existing message context.

    |Field|Type| |Description|Remarks|
    |:----|:----|:----|:----|:----|

>     |BPMCSRF|String| |The CSRF Token from the Auth response|Ex: bHZfVOIO4.iSuXwMebGibes|

Reviewer: ```suggestion
    |BPMCSRF|String| |The CSRF Token from the Auth response|e.g., bHZfVOIO4.iSuXwMebGibes|
```

>     |Set-cookie|String| |Value of Username cookie from the Auth response|Ex: UserName=83|117|112|101|114|118|105|115|111|114; expires=Sat, 07-May-2022 08:46:12 GMT; path=/; secure; HttpOnly|

Reviewer: ```suggestion
    |Set-cookie|String| |Value of Username cookie from the Auth response|e.g., UserName=83|117|112|101|114|118|105|115|111|114; expires=Sat, 07-May-2022 08:46:12 GMT; path=/; secure; HttpOnly|
```

    |Cookie|String| |Cookie Value from the Auth response| |

    **Sample configuration**

    ```xml
    <creatiocrm.init>
        <hostName>https://creatio-hostname</hostName>
        <password>creatio-username</password>
        <username>creatio-password</username>
    </creatiocrm.init>
    <creatiocrm.caseCreate/>

    ```

    **Sample request**

    ```json

    {
        "Number": "SR00000040",
        "RegisteredOn": "2021-07-16T11:00:00Z",
        "Subject": "Consultation on functionality",
        "Symptoms": "Comparison of different DBMS in accordance with the customer's needs",
        "Notes": "",
        "AccountId": "ff7e089f-1fe9-4ca9-bc30-2d76ad39d178",
        "ContactId": "00b34750-2feb-4545-b233-153502326f3c",
        "SolutionRemains": 0.0
    }

    ```

??? note "Get case"

>     The caseGet operation is used to retrieve the case already exist in the  reatioCRM with the unique id

Reviewer: ```suggestion
    The caseGet operation is used to retrieve the case already exist in the reatioCRM with the unique ID
```

>     If the given id does not exist in the CreatioCRM this will return 404 HTTP status code.

Reviewer: ```suggestion
    If the given ID does not exist in the CreatioCRM this will return the 404 HTTP status code.
```

    |Field|Type| |Description|Remarks|
    |:----|:----|:----|:----|:----|
    |BPMCSRF|String| |The CSRF Token from the Auth response|E.g., bHZfVOIO4.iSuXwMebGibes|
    |Set-cookie|String| |Value of Username cookie from the Auth response|E.g., UserName=83|117|112|101|114|118|105|115|111|114; expires=Sat, 07-May-2022 08:46:12 GMT; path=/; secure; HttpOnly|
    |Cookie|String| |Cookie Value from the Auth response| |

    **Sample configuration**

    ```xml
    <creatiocrm.init>
        <hostName>https://creatio-hostname</hostName>
        <password>creatio-username</password>
        <username>creatio-password</username>
    </creatiocrm.init>
    <creatiocrm.caseGet/>
        <id>{$ctx:uri.var.id}</id>
    </creatio.caseGet>

    ```

??? note "Update case"
    The caseUpdate operation is used to update the existing case details in the CreatioCRM. 

    If the given ID does not exist in the CreatioCRM this will create new contact with the given ID.

    |Field|Type| |Description|Remarks|
    |:----|:----|:----|:----|:----|

>     |BPMCSRF|String| |The CSRF Token from the Auth response|Ex: bHZfVOIO4.iSuXwMebGibes|

Reviewer: ```suggestion
    |BPMCSRF|String| |The CSRF Token from the Auth response|e.g., bHZfVOIO4.iSuXwMebGibes|
```

>     |Set-cookie|String| |Value of Username cookie from the Auth response|Ex: UserName=83|117|112|101|114|118|105|115|111|114; expires=Sat, 07-May-2022 08:46:12 GMT; path=/; secure; HttpOnly|

Reviewer: ```suggestion
    |Set-cookie|String| |Value of Username cookie from the Auth response|e.g., UserName=83|117|112|101|114|118|105|115|111|114; expires=Sat, 07-May-2022 08:46:12 GMT; path=/; secure; HttpOnly|
```

    |Cookie|String| |Cookie Value from the Auth response| |

    **Sample configuration**

    ```xml

    <creatiocrm.init>
        <hostName>https://creatio-hostname</hostName>
        <password>creatio-username</password>
        <username>creatio-password</username>
    </creatiocrm.init>
    <creatiocrm.caseUpdate/>
        <id>{$ctx:uri.var.id}</id>
    </creatio.caseUpdate>
    ```

    **Sample request**

    ```json

    {
        "Number": "SR00000040",
        "RegisteredOn": "2021-07-16T11:00:00Z",
        "Subject": "Consultation on functionality",
        "Symptoms": "Comparison of different DBMS in accordance with the customer's needs",
        "Notes": "",
        "AccountId": "ff7e089f-1fe9-4ca9-bc30-2d76ad39d178",
        "ContactId": "00b34750-2feb-4545-b233-153502326f3c",
        "SolutionRemains": 0.0
    }

    ```

??? note "Delete case"
    The caseDelete operation is used to delete a case that already exists in the CreatioCRM with given unique ID.

>     If the given ID does not exists in the CreatioCRM this will return 404 HTTP status code.

Reviewer: ```suggestion
    If the given ID does not exist in the CreatioCRM this will return a 404 HTTP status code.
```

    |Field|Type| |Description|Remarks|
    |:----|:----|:----|:----|:----|
    |BPMCSRF|String| |The CSRF Token from the Auth response|E.g., bHZfVOIO4.iSuXwMebGibes|
    |Set-cookie|String| |Value of Username cookie from the Auth response|E.g., UserName=83|117|112|101|114|118|105|115|111|114; expires=Sat, 07-May-2022 08:46:12 GMT; path=/; secure; HttpOnly|
    |Cookie|String| |Cookie Value from the Auth response| |

    **Sample configuration**

    ```xml

    <creatiocrm.init>
        <hostName>https://creatio-hostname</hostName>
        <password>creatio-username</password>
        <username>creatio-password</username>
    </creatiocrm.init>
    <creatiocrm.caseDelete>
        <id>{$ctx:uri.var.id}</id>
    </creatio.caseDelete>

    ```

### Account operations

??? note "Account Create"
    Creatio accountCreate operation is used in MI mediation to create a new Account with a unique ID in the Creatio. 

    The message body for the account creation will be taken from the payload of the existing message context.

    |Field|Type| |Description|Remarks|
    |:----|:----|:----|:----|:----|
    |Content-Type|String| |application/json| |

>     |BPMCSRF|String| |The CSRF Token from the Auth response|Ex: bHZfVOIO4.iSuXwMebGibes|

Reviewer: ```suggestion
    |BPMCSRF|String| |The CSRF Token from the Auth response|e.g., bHZfVOIO4.iSuXwMebGibes|
```

>     |Set-cookie|String| |Value of Username cookie from the Auth response|Ex: UserName=83|117|112|101|114|118|105|115|111|114; expires=Sat, 07-May-2022 08:46:12 GMT; path=/; secure; HttpOnly|

Reviewer: ```suggestion
    |Set-cookie|String| |Value of Username cookie from the Auth response|e.g., UserName=83|117|112|101|114|118|105|115|111|114; expires=Sat, 07-May-2022 08:46:12 GMT; path=/; secure; HttpOnly|
```

    |Cookie|String| |Cookie Value from the Auth response| |

    **Sample configuration**

    ```xml

    <creatiocrm.init>
        <hostName>https://creatio-hostname</hostName>
        <password>creatio-username</password>
        <username>creatio-password</username>
    </creatiocrm.init>
    <creatiocrm.accountCreate/> 

    ```

    **Sample request**

    ```json

    {
        "Name": "API 1",
        "ProcessListeners": 0,
        "Phone": "+1 206 480 3801",
        "AdditionalPhone": "+1 206 480 4495",
        "Fax": "",
        "Web": "www.infocom-global.com",
        "Address": "48 Pilgrim Street",
        "Notes": ""
    }

    ```

??? note "Get Account"
    The accountGet operation is used to retrieve the account that already exists in the CreatioCRM with the unique ID.

    If the given ID does not exist in the CreatioCRM, this will return a 404 HTTP status code.

    |Field|Type| |Description|Remarks|
    |:----|:----|:----|:----|:----|

>     |BPMCSRF|String| |The CSRF Token from the Auth response|Ex: bHZfVOIO4.iSuXwMebGibes|

Reviewer: ```suggestion
    |BPMCSRF|String| |The CSRF Token from the Auth response|e.g., bHZfVOIO4.iSuXwMebGibes|
```

>     |Set-cookie|String| |Value of Username cookie from the Auth response|Ex: UserName=83|117|112|101|114|118|105|115|111|114; expires=Sat, 07-May-2022 08:46:12 GMT; path=/; secure; HttpOnly|

Reviewer: ```suggestion
    |Set-cookie|String| |Value of Username cookie from the Auth response|e.g., UserName=83|117|112|101|114|118|105|115|111|114; expires=Sat, 07-May-2022 08:46:12 GMT; path=/; secure; HttpOnly|
```

    |Cookie|String| |Cookie Value from the Auth response| |

    **Sample configuration**

    ```xml
    <creatiocrm.init>
        <hostName>https://creatio-hostname</hostName>
        <password>creatio-username</password>
        <username>creatio-password</username>
    </creatiocrm.init>
    <creatiocrm.accountGet>
        <id>{$ctx:uri.var.id}</id>
    </creatio.accountGet>

    ```

??? note "Update Account"
    The accountUpdate operation is used to update the existing account details in the CreatioCRM. 

>     If the given ID does not exist in the CreatioCRM this will create new account with the given ID.

Reviewer: ```suggestion
    If the given ID does not exist in the CreatioCRM this will create a new account with the given ID.
```

    |Field|Type| |Description|Remarks|
    |:----|:----|:----|:----|:----|

>     |BPMCSRF|String| |The CSRF Token from the Auth response|Ex: bHZfVOIO4.iSuXwMebGibes|

Reviewer: ```suggestion
    |BPMCSRF|String| |The CSRF Token from the Auth response|e.g., bHZfVOIO4.iSuXwMebGibes|
```

>     |Set-cookie|String| |Value of Username cookie from the Auth response|Ex: UserName=83|117|112|101|114|118|105|115|111|114; expires=Sat, 07-May-2022 08:46:12 GMT; path=/; secure; HttpOnly|

Reviewer: ```suggestion
    |Set-cookie|String| |Value of Username cookie from the Auth response|e.g., UserName=83|117|112|101|114|118|105|115|111|114; expires=Sat, 07-May-2022 08:46:12 GMT; path=/; secure; HttpOnly|
```

    |Cookie|String| |Cookie Value from the Auth response| |

    **Sample configuration**

    ```xml
    <creatiocrm.init>
        <hostName>https://creatio-hostname</hostName>
        <password>creatio-username</password>
        <username>creatio-password</username>
    </creatiocrm.init>
    <creatiocrm.accountUpdate/>
        <id>{$ctx:uri.var.id}</id>
    </creatio.accountUpdate>

    ```

    **Sample request**

    ```json

    {
        "Name": "API 1",
        "ProcessListeners": 0,
        "Phone": "+1 206 480 3801",
        "AdditionalPhone": "+1 206 480 4495",
        "Fax": "",
        "Web": "www.infocom-global.com",
        "Address": "48 Pilgrim Street",
        "Notes": ""
    }

    ```

??? note "Delete Account"
    The accountDelete operation is used to delete an account that already exists in the CreatioCRM with a given unique ID.
    

>     If the given ID does not exist in the CreatioCRM, this will return 404 HTTP status code.

Reviewer: ```suggestion
    If the given ID does not exist in the CreatioCRM, this will return a 404 HTTP status code.
```

    |Field|Type| |Description|Remarks|
    |:----|:----|:----|:----|:----|
    |BPMCSRF|String| |The CSRF Token from the Auth response|E.g., bHZfVOIO4.iSuXwMebGibes|
    |Set-cookie|String| |Value of Username cookie from the Auth response|E.g., UserName=83|117|112|101|114|118|105|115|111|114; expires=Sat, 07-May-2022 08:46:12 GMT; path=/; secure; HttpOnly|

>     |Cookie|String| |Cookie value from the Auth response| |

Reviewer 2: The cookies will be added by the connector hence the end user doesn't need to configure them. Therefore shall we remove these `Request Header attributes` and have only `Request Parameters` as defined in the internal doc? Please check other operations as well

Reviewer 2: As per the offline discussion `Request Header attributes` should be optional. Therefore shall we mention them as optional instead of removing them? @samgnaniah
